docs/source/howto/codes.rst

```diff
@@ -4,41 +4,265 @@
+The typical way to implement these is to use ``try``/``except`` clauses to wrap the lines of code that might raise some typical errors, that can then be replaced by these ``exit_codes``:
+
+.. code-block:: python
+   :linenos:
+
+    def parse(self, **kwargs):
+        from aiida.common import exceptions
+
+        try:
+            output_folder = self.retrieved
+        except exceptions.NotExistent:
+            return self.exit_codes.ERROR_NO_RETRIEVED_FOLDER
+
+        output_filename = self.node.get_attribute('output_filename')
+
+        try:
+            with output_folder.open(output_filename, 'r') as handle:
+                try:
+                    result = int(handle.read())
+                except ValueError:
+                    result = None
+        except (OSError, IOError):
+            return self.exit_codes.ERROR_READING_OUTPUT_FILE
+
+        if result is None:
+            return self.exit_codes.ERROR_INVALID_OUTPUT
+
+        self.out('sum', orm.Int(result))
+
+You can see there is one for the case where no ``retrieve`` output was found (lines 4-7), another for the case of not being able to read the output file (lines 11-18), and finally one for when the result printed in the file is not a valid integer (lines 20-21).
+You then have to introduce all of these `exit_codes` inside of the |define| method of the |CalcJob| that will use this |Parser| (so, in that sense, when working like this the exit codes need to be "supported" by any calculation that wants to use the |Parser|).
+
+.. code-block:: python
+
+    @classmethod
+    def define(cls, spec):
+        (...)
+        spec.exit_code(300, 'ERROR_NO_RETRIEVED_FOLDER', message='The retrieved folder data node could not be accessed.')
+        spec.exit_code(310, 'ERROR_READING_OUTPUT_FILE', message='The output file could not be read from the retrieved folder.')
+        spec.exit_code(320, 'ERROR_INVALID_OUTPUT', message='The output file contains invalid output.')
+
+As you can see, for each ``exit_code`` you need to provide an ID number that will be used to identify it, a label you can then use to reference the code in the |parse| method (``self.exit_codes.LABEL``), and a message that will give the user more information on the problem.
+The Topics section on :ref:`defining processes <topics:processes:usage:defining>` also contains more information on how to use exit codes.
+
+Design guidelines
+-----------------
+
+ * | **Parse what you want to query for.**
+   | Make a list of which information to:
+
+     #. parse into the database for querying (:py:class:`~aiida.orm.nodes.data.dict.Dict`, ...)
+     #. store in the file repository for safe-keeping (:py:class:`~aiida.orm.nodes.data.singlefile.SinglefileData`, ...)
+     #. leave on the computer where the calculation ran (:py:class:`~aiida.orm.nodes.data.remote.RemoteData`, ...)
```

Comment:
It's not clear to me what the meaning of the ellipses is here.

Reply:
So, I didn't write this part so this is just a guess, but I think it is suggesting one posibility of other options? Like, to store queryable data you can do it as a `Dict` but maybe some data would be more convenient to store as an array.

Reply:
Ok, sorry the provenance of those lines got lost. Nevermind then.